services/course-material/src/components/chatbot/MessageBubble.tsx

```diff
@@ -58,33 +121,315 @@ const messageStyle = css`
+    cursor: default;
+    background-color: ${baseTheme.colors.gray[200]};
+    padding: 0 7px 0 7px;
+    border-radius: 10px;
+    font-size: 85%;
+    &:hover {
+      filter: brightness(0.9) contrast(1.1);
+    }
+    ${clicked && `box-shadow: inset 0 0 0 2px ${baseTheme.colors.gray[400]};`}
+  }
   white-space: pre-wrap;
 `
+const expandButtonStyle = css`
+  flex: 1;
+  cursor: pointer;
+  background-color: ${baseTheme.colors.gray[100]};
+  border: none;
+  margin: 0 0.5rem;
+  color: ${baseTheme.colors.gray[400]};
+  transition: filter 0.2s;
+
+  &:hover {
+    filter: brightness(0.9) contrast(1.1);
+  }
+`
+
+const referenceListStyle = (expanded: boolean, referenceList: string) => css`
+  display: flex;
+  ${expanded ? `flex-flow: column nowrap;` : `flex-flow: row nowrap;`}
+
+  #${referenceList} {
+    display: flex;
+    flex: 10;
+    ${expanded
+      ? `
+    flex-flow: column nowrap;
+    padding: 7px;
+    justify-content: space-around;
+    `
+      : `
+    flex-flow: row nowrap;
+    overflow: hidden;
+    white-space: pre;
+    mask-image: linear-gradient(0.25turn, black 66%, transparent);
+  `}
+  }
+`
 
 let md = getRemarkable()
 
 const MessageBubble: React.FC<MessageBubbleProps> = ({
   message,
   isFromChatbot,
   isPending,
-  hideCitations,
+  citations,
 }) => {
-  const processedMessage = useMemo(() => {
-    let renderedMessage = message
-    if (hideCitations) {
-      renderedMessage = renderedMessage.replace(/\[.*?\]/g, "")
-    }
+  const { t } = useTranslation()
+
+  const referenceListId = useId()
+  let triggerRef = useRef<HTMLElement>(null)
+
+  const [citationsOpen, setCitationsOpen] = useState(false)
+  const [popoverState, dispatch] = useReducer(popperStateReducer, {
+    refElement: null,
+    citationButtonClicked: false,
+  })
+
+  let { hoverProps: hoverCitationProps, isHovered: isCitationHovered } = useHover({
+    onHoverStart: (e) => {
+      dispatch({ type: "HOVER_CITATION_BUTTON", payload: e.target })
+    },
+    onHoverEnd: (e) => {
+      dispatch({ type: "UNHOVER_CITATION_BUTTON", payload: e.target })
+    },
+  })
+  let { hoverProps: hoverPopoverProps, isHovered: isPopoverHovered } = useHover({})
+
+  useEffect(() => {
+    triggerRef.current = popoverState.refElement
+  }, [popoverState.refElement])
+
+  const [processedMessage, processedCitations, citationTitleLen] = useMemo(() => {
+    let renderedMessage: ReactElement[] = []
+    let filteredCitations: ChatbotConversationMessageCitation[] = []
+
     if (isFromChatbot) {
-      renderedMessage = sanitizeCourseMaterialHtml(md.render(renderedMessage).trim())
+      let messageCopy = message
+      messageCopy = sanitizeCourseMaterialHtml(md.render(messageCopy).trim()).slice(3, -4)
+      // slicing to remove the <p>-tags that envelope the whole message, since they will be broken
+      // when the message is split into parts later
+      let citedDocs = Array.from(messageCopy.matchAll(/\[[a-z]*?([0-9]+)\]/g), (arr, _) =>
```

Comment:
Helper function that takes the text as input and returns the (citationkey, text) pairs.

https://lodash.com/docs/4.17.15#zip

Also write simple tests